### DISCUSSION TITLE
discuss whether it worked or didn't work

### DISCUSSION THREAD

**Discussion by lucidrains** [lucidrains/lion-pytorch#1] [General]
share positive or negative results here

please leave the batch size and other hyperparameters

please also leave the learning rate scheduler, if one was used

**Comment by lucidrains** [comment]
https://crumbly.medium.com/small-scale-home-evaluation-of-googles-new-optimizer-lion-77115ba8a1a4

**Comment by lucidrains** [comment]
https://github.com/lucidrains/lion-pytorch/pull/4#issuecomment-1433360309

**Comment by anhnguyen-namitech** [comment]
Didn't work for speech enhancement tasks when I ran a sweep with Random search and ASHA: optimizer=["Adam", "Adan", "Lion"] and lr is log uniform from 1e-5 to 1e-2. Note that Lion's trials were killed early in the experiments by ASHA. Here is the top performed runs:
![image](https://user-images.githubusercontent.com/103894695/219552071-09b8c21c-20da-458c-96c7-15dce3c78258.png)

I will do an extra exp without ASHA.

**Reply by lucidrains** [reply]
thanks for sharing this!

**Reply by lucidrains** [reply]
@anhnami your network, is it mostly comprised of convolutions? also, how high are your batch sizes?

**Reply by anhnguyen-namitech** [reply]
It's a UNet (20M params) with conv+attention, and bilstm at bottleneck for offline speech separation.  Batchsize = 4, each sequence is 20s long at 48 kHz (4 parallel trials, each trial uses 12GB GPU RAM because I only have 2 spare RTX3090s).  In the experiment above, I set ASHA to aggressive kill trials every 500 steps*1.5^n=1.., it seems that Lion's trials were not good in the beginning, and they were killed too early. 

I'm redoing without ASHA and with 3 lr=[2e-4, 6.6e-5, 2e-5]. Lion now looks promising, I will share the results when it's done.

**Reply by lucidrains** [reply]
@anhnami please do, thank you

**Reply by lucidrains** [reply]
try with higher batch sizes too, maybe at least 16 or 32. seems like it is batch size sensitive technique

**Reply by xiangning-chen** [reply]
Thanks for testing the Lion optimizer! We observe that at the very beginning, Lion may perform worse than Adam but it will catch up soon, there is another example: https://crumbly.medium.com/small-scale-home-evaluation-of-googles-new-optimizer-lion-77115ba8a1a4.

**Reply by anhnguyen-namitech** [reply]
Multiple runs with different seeds on a real validation set, hp_metric is predicted MOS, Lion didn't perform as good as Adam for this dense prediction task.
![image](https://user-images.githubusercontent.com/103894695/219849270-20165242-c575-4be3-a053-4d5acc6d8f71.png)

Similar results are also observed for other metrics:
![image](https://user-images.githubusercontent.com/103894695/219850381-e7cc1e45-8d71-4a57-abe5-74fb2fa636e6.png)

 I can increase the batch size, but I would need to reduce the sequence length to fit into a 3090; at the end of the day, the model will see a similar amount of data.

**Reply by lucidrains** [reply]
@anhnami  thank you!

**Comment by lucidrains** [comment]
https://twitter.com/alexjc/status/1626586985975660546

**Comment by lucidrains** [comment]
https://github.com/lucidrains/denoising-diffusion-pytorch/issues/170#issue-1589771837

**Comment by lucidrains** [comment]
this doesn't look great https://github.com/sinpcw/showcase-optimizer

**Reply by crazydonkey200** [reply]
Thanks for sharing the comparisons! I took a look at this repo and found that it applied [weight decay to AdamW](https://github.com/sinpcw/showcase-optimizer/blob/main/train.py#L55)  but [not Lion](https://github.com/sinpcw/showcase-optimizer/blob/main/train.py#L66), so I just added a pull request to fix that for fair comparison.

**Reply by lucidrains** [reply]
oh good catch! hope they update the results

personally, i think it is shaping up to be an adamw replacement for sufficiently large batch size training, based on what i'm hearing! congrats!

**Comment by lucidrains** [comment]
negative result from RL https://twitter.com/kaixhin/status/1626772629796564992

**Comment by lucidrains** [comment]
slightly worse than Adam https://twitter.com/kyo_takano/status/1627147339143200768

**Comment by lucidrains** [comment]
better than Adam at extreme batch sizes (16k) at open clip https://github.com/mlfoundations/open_clip/pull/432

**Comment by mlw214** [comment]
I'm currently investigating Lion for an object detection problem I'm working on. Unfortunately, I'm running into a lot of problems finding a good learning rate that works throughout the whole run. Seems for whatever reason it's prone to NaNing with our model/dataset when scaled up to what we used for training production models. Currently in the process of confirming if some other changes are the cause. I'll post updates here.

For reference, here's the model and a brief description of the dataset (which is proprietary):
Backbone: ConvNeXt-XL
FPN: Heavily modified BiFPN
Head: Dynamic RCNN
Current optimizer: MADGRAD
Dataset: can't say too much, but a significant portion of the objects are quite small, and for one of the classes there are a lot of confounders present. It's also fairly large.

**Reply by lucidrains** [reply]
@mlw214 thank you!

**Reply by lucidrains** [reply]
similar https://twitter.com/alexjc/status/1628646613030780935

**Reply by mlw214** [reply]
Update: taking a different approach. I'm going to get a good AdamW-based run so I can use the learning rate and weight decay values to guide my Lion tests. Just kicked off my first run. Trying to fit this in between other higher priority tasks, so things are a little slow.

**Reply by mlw214** [reply]
Ok, I got a decent AdamW run completed, though it converged slower and performed slightly worse than my best MADGRAD-based run (difference of ~0.008 mAP@50). Should be able to kick off a Lion run this weekend. I'm going to start with lr / 10, but what's the advice for the weight decay? 30x increase? @xiangning-chen

**Reply by xiangning-chen** [reply]
Thank you! Since you already got a descent AdamW run, I would say maybe first try lr / 10, wd * 10 for Lion.
I would recommend that the initial value, peak value, and end value in the learning rate schedule should be decreased simultaneously with the same ratio (10x in this case) compared to AdamW.

**Reply by xiangning-chen** [reply]
> similar https://twitter.com/alexjc/status/1628646613030780935

https://twitter.com/alexjc/status/1631535778005131269?s=20, an update is that AdamW also encounters NaNs. 
So I guess maybe this setup is a bit sensitive.

**Reply by mlw214** [reply]
It may very well be, as using even following the above values resulted in NaNs. I also tried doubling the batch size (64 to 128) and ran into the same issue. Gonna experiment making the learning "less hard" by reducing data augmentation and some other problem-specific approaches we take, including heavy use of empty imagery to reduce confounders, to see if I can't get Lion to complete a full training run.

**Reply by xiangning-chen** [reply]
@mlw214 Thanks for the update! is your task about fine-tuning a pre-trained ConvNeXt-XL on object detection task? If so, did you load the optimizer status during fine-tuning?

**Reply by mlw214** [reply]
Nope, no optimizer state is loaded. Backbone is initialized from the checkpoint Meta released (via timm), which includes only the weights afaik. Further, the FPN and detection head are trained from scratch. I'm super curious as to why Lion is having such difficulties here, so I may investigate other architectures beyond what we currently use to see if I can't narrow down the source. Note, if you have advice on approaches for debugging NaNs here I'm all ears.

**Reply by xiangning-chen** [reply]
May I know what kind of augmentations you are using? I tested with randaug and mixup, and Lion works pretty good. 
Another observation is that Lion usually leads to larger weight norm, not sure whether this affects the fine-tuning process.
I can try object detection task on my end as well.

**Reply by mlw214** [reply]
We use Trivial Augment (re-implemented using Albumentations for object detection support) with horizontal flipping, along with Mosaic and Mixup augmentations (both with p=0.5 of being applied to a given image). I'll explore in more depth some of the model watching tools Weights and Biases provides. Currently, we're tracking gradients, but not weights. I'll expand it to track weights as well to try to see if any difference between AdamW and Lion pop up.

**Reply by mlw214** [reply]
Just kicked off the following experiment:
- ConvNeXt-XL backbone
- FasterRCNN detection head
- HFlip augmentation only

If the above doesn't work I'll start experimenting with the backbone and FPN.

**Reply by mlw214** [reply]
Update: the above experiment failed. I then tried another experiment based on the above but also removed all empty images from training. That greatly helped with stability (got my longest run yet), but towards the end it still threw a NaN. Gonna swap out our custom BiFPN with the original FPN and see what happens.

**Reply by mlw214** [reply]
Alright, wasn't able to get it to work with a ConvNeXt backbone at all. However, switching to a ResNet 50 did work. I still plan to compare it against AdamW and MADGRAD, but it's a low priority at the moment since it didn't work with our production architecture, which is sad.

**Comment by lucidrains** [comment]
positive result from someone really prominent in the text-to-image field. large batch sizes (1024), learning rate / 10, weight decay kept the same

**Comment by Fred-Erik** [comment]
Tried Lion with a proprietary OCR model, roughly similar to PP-OCRv3 (https://arxiv.org/abs/2206.03001). Compared to AdamW I got much slower convergence to a lower accuracy with learning rate / 3, weight decay * 3. With the same learning rate and weight decay I get NaN results. Batch size 2048.

**Reply by lucidrains** [reply]
thanks!

**Reply by xiangning-chen** [reply]
Thanks for the experiments!
I'm not surprised that Lion got NaN with the same learning rate.
If resource is permitted, could you please also try learning rate / 10, weight decay * 10, thanks!

**Reply by Fred-Erik** [reply]
Thanks for asking; learning rate / 10, weight decay * 10 does seem to work! Experiments still going with this setup so will let you know the end results later, but it seems to converge significantly faster than AdamW.

**Reply by Fred-Erik** [reply]
![Screenshot from 2023-02-28 15-50-15](https://user-images.githubusercontent.com/2436555/221889472-a02342db-9386-4289-b585-1465ac3e3c49.png)
This experiment ended in (very slightly) lower accuracy with Lion than AdamW, but it converges much faster. This definitely has potential! Thanks for developing Lion and your follow-up question!

Green is Lion with /3,*3, blue is AdamW, orange is Lion with /10,*10. Learning rate scheduler is CosineAnnealing, but will try other learning rate schedules with Lion later.

**Reply by xiangning-chen** [reply]
Thanks for the update!
From the curve, it seems like the regularization strength is smaller in Lion.
If resource is permitted, can you please also try learning rate / 10, weight decay * 30.

Sometimes we notice that the optimal (lr*wd) for Lion differs from AdamW, although typically the difference is not significant, only around 3 times larger or smaller (see Table 12 in the paper, e.g., on Mixer-B/16, we used lr=1e-2, wd=0.3 for AdamW, while lr=3e-3, wd=3.0 for Lion). 

Also, may I know the betas settings?

As Lion is a recent development while AdamW is already well-established and well-tuned, I would greatly appreciate your help in gathering more data points. Thank you!

**Reply by Fred-Erik** [reply]
Sure! My follow-up experiments will have to wait as there is more urgent need for the available compute, but will let you know the results when I get to it.

Meanwhile, my beta settings are 0.9 and 0.99. AdamW learning rate is 1e-3, weight decay 3e-1. So with your suggested settings they would be lr=1e-2 and wd=9.0. Cosine Annealing LR schedule goes from LR to 1e-5.

**Reply by xiangning-chen** [reply]
Thank you!

I think for Lion it should be lr=1e-4, wd=9.0 right?

Another issue is the final learning rate, can you also make it 10x smaller for Lion, i.e., 1e-6.
A small learning rate at the end of training would help with the convergence to more generalizable areas.

**Reply by Fred-Erik** [reply]
Haha of course you're right. And thank you, that's very helpful. Will try it!

**Reply by lucidrains** [reply]
@xiangning-chen once (if) you can resolve some of these issues, it would be nice for you to update the paper with some of these details (say, learning rate cooldown), to increase the chances of success for those trying lion

**Reply by Fred-Erik** [reply]
Did some more experiments. wd=9.0 converges much slower, so weight decay * 30 seems to be too much. But also dividing the ending learning rate of the cosine learning rate schedule does work! So with lr / 10, wd * 10, end lr / 10, Lion is better than best settings for AdamW (green is Lion):

![image](https://user-images.githubusercontent.com/2436555/223687615-71d1e8ef-eeae-412e-8456-2211ce5cf2c0.png)

Thank you for your help!

**Reply by lucidrains** [reply]
@Fred-Erik thank you!

**Reply by xiangning-chen** [reply]
@Fred-Erik Thank you so much!

**Reply by Fred-Erik** [reply]
Actually just did a new observation: post-training quantization consistently does not work with a Lion-optimized network, where it does work with an AdamW-trained network (n=4). I tried dynamic and static quantization (using ONNXRuntime: https://onnxruntime.ai/docs/performance/model-optimizations/quantization.html) and both have the same issue of nonsensical output. An analysis of the difference in weights and activations per layer reveals there is no one specific layer to blame, but lots of layers have a much greater difference between the float32 output and int8 when trained with Lion than when trained with AdamW. So for a version of my model that needs quantization, I'm reverting back to AdamW.

**Comment by lucidrains** [comment]
negative result for LM https://twitter.com/VHellendoorn/status/1630737104975085568

edit: now positive result! https://twitter.com/VHellendoorn/status/1631349009473478656?s=20

**Reply by xiangning-chen** [reply]
They are using the same weight decay for Lion and Adam, while a 10x smaller learning rate for Lion. Hope they can update the results.

**Reply by lucidrains** [reply]
i don't think that's true, if you look at his follow up tweet.

**Reply by xiangning-chen** [reply]
https://twitter.com/VHellendoorn/status/1631097679693463554?s=20

**Reply by lucidrains** [reply]
> https://twitter.com/VHellendoorn/status/1631097679693463554?s=20

! 🤞

**Reply by xiangning-chen** [reply]
positive now: https://twitter.com/VHellendoorn/status/1631349009473478656?s=20

**Reply by lucidrains** [reply]
@xiangning-chen amazing! thank you! even if Lion is only useful for language model training, that is a huge step forward

**Reply by lucidrains** [reply]
i place a lot of weight on Vincent's results, things are looking upwards!

**Comment by lucidrains** [comment]
positive result, 3x faster for vision transformer fine tuning https://twitter.com/Haoxiang__Wang/status/1631355469439590412?s=20

**Comment by lucidrains** [comment]
yet another positive result for training LLM from a really good researcher, however he also told me fine tuning was not as good (albeit the based model was trained with Adam, not sure if that makes any difference)

**Reply by xiangning-chen** [reply]
Thanks for the update! I was wondering, when fine-tuning, do they load the Adam optimizer state from pre-training?
Previously we find that this is helpful. However, as the LLM is pre-trained by Adam, Lion is unable to utilize the optimizer state. Therefore, to ensure a fair comparison, we disable this feature for both optimizers.

**Reply by lucidrains** [reply]
@xiangning-chen oh yes, that's a good point, i'll ask him about that next time we chat

**Comment by anhnguyen-namitech** [comment]
Saw this, could be relevant to Lion: https://openreview.net/pdf?id=a65YK0cqH8g

**Comment by Kingsleyandher** [comment]
Hello, I used megatron to train GPT-2 with 16B parameter number, but when the learning rate increased to 3e-6, the gradient overflow led to the failure of convergence. May I ask why?

My hyperparameter：
Adamw: lr=1.5e-4 min-lr=1e-5 wd=0.01
Lion: lr=1.5e-5 min-lr=1e-6 wd=0.1 beta1=0.95 beta2-0.98

**Reply by Kingsleyandher** [reply]
beta2=0.98

**Reply by Kingsleyandher** [reply]
We used the same configuration to achieve results consistent with Adamw as baseline on the 4B parameter model

**Reply by Kingsleyandher** [reply]
micro batch size: 4 
global batic size: 32 
lr_decay_style: cosine
warmup steps: 3200

**Reply by xiangning-chen** [reply]
Hi, thanks for trying out!

Are you referring to when the learning rate increases to 3e-6 in the warmup phase, the model training explodes?
Have you encountered a similar issue with convergence while employing AdamW to train the 16B model as well?

**Comment by lapp0** [comment]
Seeing much better results in my `gpt2` distillation test with lion than any other optimizer. Loss function is KL divergence of logits and hidden states between student model and teacher model. Uses default parameters [here](https://github.com/lapp0/distily/blob/61c78e3136b76d5c787b8efc49125dc70dc04004/distily/args.py) only adjusting the optimizer

 `lion_32bit` > `lion_8bit` > `adamw_8bit` ≈ `adamw_32bit` > all other optimizers

Results:

https://huggingface.co/lapp0/distily_bench_gpt2_optim_extended2/tensorboard

![image](https://github.com/user-attachments/assets/923f7eab-f4d9-4a95-a7b6-09ba9f4753ac)
